### ISSUE TITLE
tensor_regression_layer_pytorch.ipynb | penalty term?

### ISSUE THREAD

**Issue by Audrius-St** [JeanKossaifi/tensorly-notebooks#8]
Hello, 

Thank you for providing the tensor_regression_layer_pytorch.ipynb example. Most helpful in understanding your paper. Ran the example successfully, reproducing the result

> Train Epoch: 19 [0/60000 (0%)]  Loss: 0.006197
> Train Epoch: 19 [16000/60000 (27%)]     Loss: 0.006317
> Train Epoch: 19 [32000/60000 (53%)]     Loss: 0.006328
> Train Epoch: 19 [48000/60000 (80%)]     Loss: 0.006382
>  mean: 1.0914269488182526e-09
> 
> Test set: Average loss: 0.0000, Accuracy: 9894/10000 (98%)

However, noticed that in the _TRL_ class, in the member function _penalty_
```
    def penalty(self, order=2):
        penalty = tl.norm(self.core, order)
        for f in self.factors:
            penatly = penalty + tl.norm(f, order)
        return penalty
```
_penalty_ has been inadvertently misspelled as  _penatly_  in the line       

`penatly = penalty + tl.norm(f, order)`

which means that the factor penalties are not being included in the penalty sum.

Corrected this and reran the example, which yielded  the puzzling result

> Train Epoch: 19 [0/60000 (0%)]  Loss: 2.302585
> Train Epoch: 19 [16000/60000 (27%)]     Loss: 2.302585
> Train Epoch: 19 [32000/60000 (53%)]     Loss: 2.302585
> Train Epoch: 19 [48000/60000 (80%)]     Loss: 2.302585
> mean: 0.00023025853442959487
> 
> Test set: Average loss: 0.0002, Accuracy: 980/10000 (9%)

No improvement in the accuracy over the 19 epochs of the run.

In _Section 4.1. Implementation_ 
of your paper, it is stated that

> In addition we constrain the weights of the tensor regression by applying L2 normalization (Salimans & Kingma, 2016) to the factors of the Tucker decomposition.

At a bit of a loss trying to understand and reconcile what is written in the paper and what is observed in the example. Any insight would be appreciated.

Thank you.

**Comment by Audrius-St** [comment]
The following modifications address the issue, at least in the example CNN.

Replaced the L2 norm penalty regularization with spectral normalization of the regression weights tensor.

```
class TRL(nn.Module):
    def __init__(self,
                 input_size,
                 ranks,
                 n_modes_b, # n_modes_b = 1 in the original code; a generalization
                 output_size,
                 verbose=1,
                 **kwargs):
        super(TRL, self).__init__(**kwargs)
        self.ranks = list(ranks)
        self.n_modes_b = n_modes_b
        self.verbose = verbose

        if isinstance(output_size, int):
            self.input_size = [input_size]
        else:
            self.input_size = list(input_size)

        if isinstance(output_size, int):
            self.output_size = [output_size]
        else:
            self.output_size = list(output_size)

        self.n_outputs = int(np.prod(output_size[self.n_modes_b:]))

        # Core of the regression tensor weights
        self.core = nn.Parameter(tl.zeros(self.ranks), requires_grad=False)
        self.bias = nn.Parameter(tl.zeros(1), requires_grad=True)
        weight_size = (list(self.input_size[self.n_modes_b:]) +
                       list(self.output_size[self.n_modes_b:]))

        # Add and register the factors
        self.factors = []
        for index, (in_size, rank) in enumerate(zip(weight_size, ranks)):
            self.factors.append(nn.Parameter(tl.zeros((in_size, rank)),
                                requires_grad=False))
            self.register_parameter('factor_{}'.format(index),
                                    self.factors[index])

        # FIX THIS
        self.core.data.uniform_(-0.1, 0.1)
        for f in self.factors:
            f.data.uniform_(-0.1, 0.1)

        # Register the regression weights tensor as a parameter
        self.regression_weights = nn.Parameter(
                                        tl.tucker_to_tensor(self.core,
                                                            self.factors),
                                        requires_grad=True)
        self.register_parameter('regression_weights', self.regression_weights)

    def forward(self, x):
        # print('x', x.shape)
        tr = inner(
                x,
                self.regression_weights,
                n_modes=tl.ndim(x) - self.n_modes_b) + self.bias
        # print('tr', tr.shape, '\n')
        return tr
```

In 

`class Net(nn.Module):`

the TRL is now instantiated as

```
        self.trl3 = nn.utils.spectral_norm(
                        TRL(input_size=(batch_size, 50, 4, 4),
                            ranks=(10, 3, 3, 10),
                            n_modes_b=1,
                            output_size=(batch_size, 10)),
                        name='regression_weights')
```

Removing the regularization penalty term,  the loss function simplifies to

`loss = criterion(output, target)`

With batch normalization, one now obtains

> Train Epoch: 19 [0/60000 (0%)]  Loss: 0.053898
> Train Epoch: 19 [16000/60000 (27%)]     Loss: 0.076959
> Train Epoch: 19 [32000/60000 (53%)]     Loss: 0.000862
> Train Epoch: 19 [48000/60000 (80%)]     Loss: 0.012166
> mean: 3.311940872663399e-07
> 
> Test set: Average loss: 0.0000, Accuracy: 9906/10000 (99%)

**Comment by JeanKossaifi** [comment]
Thanks for spotting the issue with the penalty. The TRL here is written to be an easy /accessible tutorial, not to reproduce the results of the paper (we report results on ImageNet, not MNIST). In particular, I added the penalty simply to demonstrate the concept. In practice, you already would have a non-zero weight decay that would act as an l2 regularizer on the weights. Note that in the paper we do not use spectral norm but l2 normalisation so you'd want weight_norm rather than spectral norm.
